### ISSUE TITLE
Reducing state witness size

### ISSUE THREAD

**Issue by pugachAG** [near/nearcore#10780]
We are interested in optimizing 2 parameters:
- Max state witness size. Having large state witness size could result in long distribution time which could result in chunk endorsement not being sent to block producer in time. This is partially addressed by https://github.com/near/nearcore/issues/10259. 
- Total size (the save as avg size). Outbound network traffic is expensive, so we want to reduce it as much as possible.

One important observation is that majority of state witness size is occupied by values with relatively large size (128KB+): [dashboard](https://nearinc.grafana.net/d/a41c3c5e-4c1b-41fc-919c-11040e9cfc13/shadow-chunk-validation?orgId=1&from=1710336658780&to=1710337979128&var-node_id=pugachag-mainnet&var-shard_id=All&viewPanel=26). Those values correspond to `ContractCode`.

### Compression

One straightforward way to reduce state witness size is to compress it before sending it to validators. The expectation is that wasm contract code should be compression-friendly.
One downside is that using compression adds non-trivial latency to encode + decode data.
We evaluated zstd compression with different levels: 2, 3, 5, 10.

### Caching large state values

The idea is for stateless validators to have some of the large state values saved on disk. Chunk producers know which values are present for each validator and don't include those as part state witness storage proofs. When validating a chunk using state witness those values will be read from disk. Additional disk read latency should not be an issue since we only do that for a small subset of relatively big values.

High level algorithm description:
- Stateless validators can have up to 1000 state values present. They maintain LRU cache of value hashes per shard. Each time this set changes, the validator sends a new set to all chunk producers for current and next epochs.
- Before sending a state witness chunk producer removes all values that are present in the set for that validator.
This can be easily emulated as part of shadow chunk validation.

### Results

- State witness size per second: [dashboard](https://nearinc.grafana.net/d/a41c3c5e-4c1b-41fc-919c-11040e9cfc13/shadow-chunk-validation?orgId=1&from=now-1h&to=now&var-node_id=pugachag-mainnet&var-shard_id=All&viewPanel=28)
- Max state witness size: [dashbpard](https://nearinc.grafana.net/d/a41c3c5e-4c1b-41fc-919c-11040e9cfc13/shadow-chunk-validation?orgId=1&from=now-1h&to=now&var-node_id=pugachag-mainnet&var-shard_id=All&viewPanel=29)
- Compression time: [dashboard](https://nearinc.grafana.net/d/a41c3c5e-4c1b-41fc-919c-11040e9cfc13/shadow-chunk-validation?orgId=1&from=now-1h&to=now&var-node_id=pugachag-mainnet&var-shard_id=All&viewPanel=30)

**Comment by walnut-the-cat** [comment]
Thanks for the write up-! couple questions
1. Could you share how I can read [dashboard](https://nearinc.grafana.net/d/a41c3c5e-4c1b-41fc-919c-11040e9cfc13/shadow-chunk-validation?orgId=1&from=1710336658780&to=1710337979128&var-node_id=pugachag-mainnet&var-shard_id=All&viewPanel=26) better? 
2. What does 'cache_state_values' in witness size graphs mean? Does it mean size of state witness when 'we don't do any compression but enable state value caching on validator side'?
3. How long does 'decompression' take? Do we expect similar amount time to compression time? (thus 2x compression time latency)

**Comment by pugachAG** [comment]
@walnut-the-cat 
1. I've added more detailed description for that dashboard
2. Yes, it means caching state values with no compression
3. decompression time is already included, added that to the description

**Comment by pugachAG** [comment]
@walnut-the-cat yes, #10566 is pretty much the same as "Caching large state values" but with a slightly different approach

**Comment by walnut-the-cat** [comment]
Then can we merge #10566 into this one?

**Comment by shreyan-gupta** [comment]
> Top N values as part of the Protocol

It was mentioned that we would globally cache to top N contracts based on usage. A good metric to use here can be `size_of_contract * num_times_called_in_prev_epoch`? This would provide the maximum size benefits assuming the load of the current epoch would be similar to the previous epoch.

**Comment by pugachAG** [comment]
@shreyan-gupta `size_of_contract * num_times_called_in_prev_epoch` makes sense if we have a lot of cache churn. In my experiments with `N=1000` per each shard it didn't fill even half-way in 6 days: [dashboard](https://nearinc.grafana.net/d/a41c3c5e-4c1b-41fc-919c-11040e9cfc13/shadow-chunk-validation?orgId=1&from=1710455340012&to=1711006671918&var-node_id=pugachag-mainnet&var-shard_id=All&viewPanel=20). So it seems like straight-forward LRU would work perfectly fine.
